### PR TITLE
stats: add integration test to flag missing tag extraction rules

### PR DESCRIPTION
**Pull request by ggreenway**
Signed-off-by: Greg Greenway <ggreenway@apple.com>

Commit Message: This is a rough heuristic, but may catch many cases of missing rules.
Additional Description:
Risk Level: low: test only
Testing: this is a new test
Docs Changes: not needed
Release Notes: not needed
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by ggreenway** [comment]
@jmarantz after seeing many new stats added without correct tag extraction rules, I wrote this. 

Pros:
* Gives some awareness to developers if they've missed this, for the most obvious cases

Cons:
* Adds some code to every single integration test. Will this cause a measurable degradation in build/test time?
* Misses anything that doesn't use `stat_prefix` in the config
* Misses anything that isn't configured in an integration test

What do you think? Is there a better way to do this, that is likely to be hit for new code with a developer who doesn't know about the tag extraction rules?

File: test/integration/base_integration_test.cc
Comment (jmarantz) [inline test/integration/base_integration_test.cc:709]: where does this magic `"stat_prefix"` come from? Could this be a string-constant?

File: test/integration/base_integration_test.cc
Comment (jmarantz) [inline test/integration/base_integration_test.cc:737]: nit; could use test_server_->forEachCounter here and void making the intermediate array. Not sure if that would be material.

**Review by jmarantz — COMMENTED** [review]
code seems fine ; I wasn't aware of this `stat_prefix` convention (or had forgotten) and so I think I'm not understanding the premise of what you are checking.

File: test/integration/base_integration_test.cc
Comment (ggreenway) [reply test/integration/base_integration_test.cc:709]: `stat_prefix` is the conventional name in the config protos.

```
$ grep -R stat_prefix api/envoy/*
api/envoy/api/v2/core/grpc_service.proto:    string stat_prefix = 4 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/listener/v3/listener.proto:  // `listener.<stat_prefix>.`.
api/envoy/config/listener/v3/listener.proto:  string stat_prefix = 28;
api/envoy/config/core/v3/grpc_service.proto:    string stat_prefix = 4 [(validate.rules).string = {min_len: 1}];
api/envoy/config/filter/udp/udp_proxy/v2alpha/udp_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/rbac/v2/rbac.proto:  string stat_prefix = 3 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/zookeeper_proxy/v1alpha1/zookeeper_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:  string stat_prefix = 2 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/dubbo_proxy/v2alpha1/dubbo_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/ext_authz/v2/ext_authz.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/local_rate_limit/v2alpha/local_rate_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/kafka_broker/v2alpha1/kafka_broker.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/rate_limit/v2/rate_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/client_ssl_auth/v2/client_ssl_auth.proto:  string stat_prefix = 2 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/mysql_proxy/v1alpha1/mysql_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/config/filter/network/thrift_proxy/v2alpha1/thrift_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_bytes: 1}];
api/envoy/extensions/matching/input_matchers/ip/v3/ip.proto:  string stat_prefix = 2 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/udp/dns_filter/v3/dns_filter.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/rbac/v3/rbac.proto:  string shadow_rules_stat_prefix = 5;
api/envoy/extensions/filters/network/rbac/v3/rbac.proto:  string stat_prefix = 3 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/local_ratelimit/v3/local_rate_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/connection_limit/v3/connection_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/zookeeper_proxy/v3/zookeeper_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:  string stat_prefix = 2 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/dubbo_proxy/v3/dubbo_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/ext_authz/v3/ext_authz.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/ratelimit/v3/rate_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/client_ssl_auth/v3/client_ssl_auth.proto:  string stat_prefix = 2 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/network/thrift_proxy/v3/thrift_proxy.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/http/rbac/v3/rbac.proto:  string shadow_rules_stat_prefix = 3;
api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/http/compressor/v3/compressor.proto:  //    `<stat_prefix>.compressor.<compressor_library.name>.<compressor_library_stat_prefix>.response.*`
api/envoy/extensions/filters/http/compressor/v3/compressor.proto:  //    `<stat_prefix>.compressor.<compressor_library.name>.<compressor_library_stat_prefix>.*`.
api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:  //         stat_prefix: waf # This emits ext_authz.waf.ok, ext_authz.waf.denied, etc.
api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:  //         stat_prefix: blocker # This emits ext_authz.blocker.ok, ext_authz.blocker.denied, etc.
api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto:  string stat_prefix = 13;
api/envoy/extensions/filters/http/bandwidth_limit/v3/bandwidth_limit.proto:  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:  string stat_prefix = 8;
```

File: test/integration/base_integration_test.cc
Comment (jmarantz) [inline test/integration/base_integration_test.cc:731]: I don't understand how just doing a total substring search for the prefixes in the tag-extracted name tells you whether the tag-extraction regexes are complete.

Comment (ggreenway) [reply test/integration/base_integration_test.cc:731]: The idea is that if the `stat_prefix` is in the tag-extracted name, then it wasn't extracted, which I believe is always an error (although there are many such cases in the codebase right now). 

Or at least that's how I use `stat_prefix`. Maybe others use it completely differently, and this isn't correct. I'd like to hear opinions on this.

Comment (ggreenway) [reply test/integration/base_integration_test.cc:731]: Also, this assumes that the configured `stat_prefix` in the integration tests is a value that is unique enough that it isn't  a substring of any other part of the stat name.

Comment (ggreenway) [reply test/integration/base_integration_test.cc:731]: Another approach to this test, that may be more robust, is to instead look at only the extracted tags, and verify that the value configured for `stat_prefix` is in one of the extracted tags, and ignore the tagExtractedName. That may more clearly express what I'm trying to accomplish.

**Comment by jmarantz** [comment]
Greg should I review now or wait till CI is clean?

File: test/integration/base_integration_test.cc
Comment (jmarantz) [inline test/integration/base_integration_test.cc:752]: you didn't want to iterate over the tags instead? Maybe explain why in a code comment?

**Review by jmarantz — APPROVED** [review]
lgtm basically modulo minor nits.

Does it make sense to just get one other maintainer's opinion on this plan generally? I'll leave it up to you on all counts.

File: test/extensions/access_loggers/open_telemetry/access_log_integration_test.cc
Comment (ggreenway) [reply test/extensions/access_loggers/open_telemetry/access_log_integration_test.cc:54]: I believe this is correct. A bunch of tests with outgoing grpc sideband connections have stats in the `grpc.` namespace.

File: test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc
Comment (mattklein123) [inline test/extensions/transport_sockets/tls/integration/ssl_integration_test.cc:896]: Can you expand? Unclear how. Same elsewhere.

File: test/integration/base_integration_test.cc
Comment (mattklein123) [reply test/integration/base_integration_test.cc:735]: Maybe add a comment here that this is based on precedent but isn't guaranteed to be exhaustive? We might need to expand this in the future?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this is awesome. I left some small comments. Can you also open a tracking beginner issue for fixing the missing extractions? Seems like a great one for new contributors.

/wait

File: test/integration/base_integration_test.cc
Comment (ggreenway) [reply test/integration/base_integration_test.cc]: I removed it; I think I must have made later changes that removed the need for it (probably disabling a specific test).

**Review by mattklein123 — APPROVED** [review]
Nice!

**Comment by alyssawilk** [comment]
hey @ggreenway @jmarantz can we have this off by default or maybe only on for CI?  It makes it significantly more complicated to debug integration tests when there's "spurious" e2e requests hitting the config dump URL

**Comment by jmarantz** [comment]
I'm not clear on why this would cause debug complexity but running it in only one CI seems OK to me.

**Comment by alyssawilk** [comment]
as the person who debugs most of our integration test flakes, it means for every test there's a ton of spurious data plane work going on which makes it significantly harder to debug what's actually broken.  I've noticed this a number of times helping people who got confused on test debug and was inspired to root case recently.

**Comment by jmarantz** [comment]
this makes sense. Thanks Alyssa.

**Pull request merged**
